Este codigo toma las features calculadas como entrada y le aplica un preprocesamiento. Se desea ir en la linea de que crear pipelines.

In [1]:
from google.colab import drive
drive.mount('/content/drive')
%cd "/content/drive/My Drive/repo_tesis/entorno_tesis_Molina"
!source bin/activate

Mounted at /content/drive
/content/drive/My Drive/repo_tesis/entorno_tesis_Molina


In [2]:
import matplotlib.pyplot as plt
import numpy as np
import joblib
import sys
import pandas as pd
import plotly.express as px

sys.path.append('/content/drive/My Drive/repo_tesis/lib')


In [76]:
features = joblib.load("/content/drive/My Drive/repo_tesis/data/FEATURES_W200_I50.joblib")
label = joblib.load("/content/drive/My Drive/repo_tesis/data/label_W200_I50.joblib")

In [6]:
print(features.info())
print(features.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64526 entries, 0 to 64525
Columns: 240 entries, wl_c1 to DWT3_c16
dtypes: float64(240)
memory usage: 118.2 MB
None
      wl_c1     zc_c1    ssc_c1    rms_c1    mav_c1     ls_c1    mfl_c1  \
0  0.070329  0.571429  0.571429  0.070710  0.068562  0.067285  0.381483   
1  0.067801  0.542857  0.514286  0.075924  0.070908  0.070481  0.387028   
2  0.080441  0.571429  0.571429  0.085491  0.085498  0.084059  0.416062   
3  0.086417  0.600000  0.685714  0.082098  0.084725  0.081804  0.419296   
4  0.090094  0.600000  0.742857  0.086324  0.089802  0.086698  0.434145   

     msr_c1   wamp_c1    iav_c1  ...    ls_c16   mfl_c16   msr_c16  wamp_c16  \
0  0.265203  0.948718  0.068372  ...  0.027815  0.165989  0.132344  0.769231   
1  0.266386  0.923077  0.068570  ...  0.035638  0.250961  0.152028  0.743590   
2  0.299510  0.948718  0.080603  ...  0.037588  0.259464  0.158806  0.692308   
3  0.298214  0.974359  0.085612  ...  0.044048  0.283299  0.1876

Se puede ver que la dimension de los datos de entrada es 240. Por cada feature, se tienen 16 canales, ya que la adquisicion se hizo con 16 electrodos.

In [ ]:
# # Bloque para ejecutar si se esta trabajando deteccion gesto/no gesto
# print(y_train[39000:40000])
# y_train[y_train!=0] = 1
# print(y_train[39000:40000])

# print(y_val[39000:40000])
# y_val[y_val!=0] = 1
# print(y_val[39000:40000])


In [85]:
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

scaler = StandardScaler()
scaler.fit(features)
data_scaled = scaler.transform(features)

pca = PCA(n_components=0.95)
data_red = pca.fit_transform(data_scaled)

print('dim antes PCA: ' + str(data_scaled.shape[1]))
print('dim luego PCA: ' + str(data_red.shape[1]))
print(pca.explained_variance_ratio_)
print(data_red.shape)

dim antes PCA: 240
dim luego PCA: 44
[0.48222788 0.0707715  0.05574917 0.04831911 0.03577393 0.03060899
 0.02225433 0.01859724 0.01602055 0.01196925 0.01133304 0.0104399
 0.00899097 0.00825947 0.00800077 0.00723603 0.0068686  0.00637792
 0.00600685 0.00579654 0.00562212 0.00514768 0.00492524 0.00459901
 0.00453724 0.00424751 0.00419127 0.0039471  0.0038074  0.00371698
 0.0035398  0.00341392 0.00322093 0.00301993 0.00290953 0.00259785
 0.00250516 0.00232054 0.00207587 0.00194695 0.0019048  0.00175597
 0.00165218 0.00155356]
(64526, 44)


In [91]:
# Guardar nuevos datos
%cd "/content/drive/My Drive/repo_tesis/data"
data_red = pd.DataFrame(data_red)
joblib.dump(data_red, "../data/FEATURES_W200_I50_PCA95.joblib")
joblib.dump(label, "../data/label_W200_I50.joblib")

/content/drive/My Drive/repo_tesis/data


['../data/label_W200_I50.joblib']

Visualización de características

In [86]:
label = np.array(label)
vector_gestos = label[:, 1]
ind_gesto1 = np.where(vector_gestos==1)[0]
ind_gesto2 = np.where(vector_gestos==2)[0]
ind_gesto3 = np.where(vector_gestos==3)[0]
ind_gesto4 = np.where(vector_gestos==4)[0]
ind_gesto5 = np.where(vector_gestos==5)[0]
ind_gesto6 = np.where(vector_gestos==6)[0]
ind_gesto7 = np.where(vector_gestos==7)[0]
ind_gesto8 = np.where(vector_gestos==8)[0]
ind_gesto9 = np.where(vector_gestos==9)[0]


In [87]:
num_feature1 = 0
num_feature2 = 1
num_feature3 =2

x1 = data_red[ind_gesto1, num_feature1]
y1 = data_red[ind_gesto1, num_feature2]
z1 = data_red[ind_gesto1, num_feature3]

x2 = data_red[ind_gesto2, num_feature1]
y2 = data_red[ind_gesto2, num_feature2]
z2 = data_red[ind_gesto2, num_feature3]

x3 = data_red[ind_gesto3, num_feature1]
y3 = data_red[ind_gesto3, num_feature2]
z3 = data_red[ind_gesto3, num_feature3]

x4 = data_red[ind_gesto4, num_feature1]
y4 = data_red[ind_gesto4, num_feature2]
z4 = data_red[ind_gesto4, num_feature3]

x5 = data_red[ind_gesto4, num_feature1]
y5 = data_red[ind_gesto4, num_feature2]
z5 = data_red[ind_gesto4, num_feature3]

x6 = data_red[ind_gesto5, num_feature1]
y6 = data_red[ind_gesto5, num_feature2]
z6 = data_red[ind_gesto5, num_feature3]

# Truncar los datos para que tengan la misma longitud
min_length = min(len(x1), len(x2), len(x3))
x1_truncated, y1_truncated, z1_truncated = x1[:min_length], y1[:min_length], z1[:min_length]
x2_truncated, y2_truncated, z2_truncated = x2[:min_length], y2[:min_length], z2[:min_length]
x3_truncated, y3_truncated, z3_truncated = x3[:min_length], y3[:min_length], z3[:min_length]
x4_truncated, y4_truncated, z4_truncated = x4[:min_length], y4[:min_length], z4[:min_length]
x5_truncated, y5_truncated, z5_truncated = x5[:min_length], y5[:min_length], z5[:min_length]
x6_truncated, y6_truncated, z6_truncated = x6[:min_length], y6[:min_length], z6[:min_length]

# Lista de nombres personalizados para los conjuntos de datos
data_names = {'red': 'Clase 1', 'green': 'Clase 2', 'blue': 'Clase 3','lime': 'Clase 4', 'black': 'Clase 5', 'grey': 'Clase 6'}

# Combinar datos en una lista de listas
combined_data = [[x1_truncated[i], y1_truncated[i], z1_truncated[i], 'red'] for i in range(min_length)]
combined_data.extend([[x2_truncated[i], y2_truncated[i], z2_truncated[i], 'green'] for i in range(min_length)])
combined_data.extend([[x3_truncated[i], y3_truncated[i], z3_truncated[i], 'blue'] for i in range(min_length)])
combined_data.extend([[x4_truncated[i], y4_truncated[i], z4_truncated[i], 'lime'] for i in range(min_length)])
combined_data.extend([[x5_truncated[i], y5_truncated[i], z5_truncated[i], 'black'] for i in range(min_length)])
combined_data.extend([[x6_truncated[i], y6_truncated[i], z6_truncated[i], 'black'] for i in range(min_length)])



# Crear DataFrame
df = pd.DataFrame(combined_data, columns=['x', 'y', 'z', 'colors'])

# Crear el scatter plot 3D
fig = px.scatter_3d(df, x='x', y='y', z='z', color='colors',
                    title='Características en 3 dimensiones')

# Personalizar la leyenda
fig.for_each_trace(lambda t: t.update(name=data_names[t.name]))

# Mostrar el scatter plot
fig.show()